Issue/7959 glide avatar upload

malinajirka opened the pull request:
Fixes #7959 

Replaces Volley with Glide in Me and SignUpEpilogue sections. 

Gravatar URLs are static -> changing the image doesn’t change the URL. When we upload a new Gravatar, the URL doesn’t change and it takes several seconds before the Gravatar server starts returning the new Gravatar image (there is a replication lag). In other words, after the upload of the new image completes, the Gravatar server still keeps returning the old Gravatar.

This solution is more of a hack/workaround than a real solution. It's not bullet proof, but works in most cases. As soon as the replication lag is fixed on the sever side we can remove it (not sure if it ever happens).

The solution has 2 parts
1. When the upload completes we load the image file from the disc into the ImageView and we save it's Bitmap into a local memory cache. As long as the bitmap is in the cache, we don't hit the server nor Glide's cache.

2. We use [Glide's signatures](https://bumptech.github.io/glide/doc/caching.html#custom-cache-invalidation) to invalidate Glide's cache.

This approach doesn't work if the user restarts the app right after they upload an image, because after the app is restarted Glide hits the server and stores the image into it's cache even though the Gravatar API returned the old image.

To test:
Sign up epilogue
1. Sign up to WordPress
2. Click on the avatar image on the Sign up Epilogue screen 
3. Upload a new avatar
4. Rotate your device
5. Click on Continue
6. Click on Me section
7. Click on the avatar image
8. Upload a new avatar

daniloercoli commented:
Nice work @malinajirka - I left some comments below, it's up to you to fix those in this PR or open new tickets for them, since I don't know it related to your update.

- I noticed that there was no loading indicator the first time I updated the `Me` picture on a brand new account. See the video here: https://cloudup.com/c-E0xbj18VV (No loading indicator - At the end of the video the picture is update correctly).

- During my tests I also noticed that the same problem could happen if you're updating the current picture https://cloudup.com/cP5cGAc5Nkl 
This is not happening all the times.

- The error message about "Failed to upload" has a white background and I barely noticed it when happened, considering that uploading took some time and I lost attention on the screen.
![screen shot 2018-10-11 at 12 43 38](https://user-images.githubusercontent.com/518232/46798877-76911780-cd53-11e8-8854-bb586d402a2c.png)

It seems to me that Gravatar servers take lot of time to update the picture. Do we know how long it takes before the replica is done?

malinajirka commented:
Thanks @daniloercoli  for the review! ;)

>I noticed that there was no loading indicator the first time I updated the Me picture on a brand new account. See the video here: https://cloudup.com/c-E0xbj18VV (No loading indicator - At the end of the video the picture is update correctly).
During my tests I also noticed that the same problem could happen if you're updating the current picture https://cloudup.com/cP5cGAc5Nkl
This is not happening all the times.

Fixed in [fd1808c](https://github.com/wordpress-mobile/WordPress-Android/pull/8409/commits/fd1808cba0ea37b5ef21f52edf894c87c14b73e6).

>The error message about "Failed to upload" has a white background and I barely noticed it when happened, considering that uploading took some time and I lost attention on the screen.

It's a regular toast message -> I'm not sure there is an easy fix for this issue. It's derived from the color theme of your device. Wdyt?

>It seems to me that Gravatar servers take lot of time to update the picture. Do we know how long it takes before the replica is done?

The replica usually takes from 5 - 30 s. However, we are not waiting for the replication - we are showing the progress indicator while the upload is in progress.

daniloercoli commented:
Hey @malinajirka - I see travis unhappy with this PR. 
I restarted it 3 times, since `./gradlew -PdisablePreDex checkstyle ktlint assembleVanillaRelease lintVanillaRelease testVanillaRelease || (grep -A20 -B2 'severity="Error"' -r --include="*.xml" WordPress libs; exit 1)` locally on my laptop does end with success, but Travis doesn't turn green.

Would you mind to check if it's a problem on Travis, or in your PR?

daniloercoli commented:
>>The error message about "Failed to upload" has a white background and I barely noticed it when happened, considering that uploading took some time and I lost attention on the screen.

>It's a regular toast message -> I'm not sure there is an easy fix for this issue. It's derived from the color theme of your device. Wdyt?

Maybe @iamthomasbishop and/or @theck13 can help on this.
It was a stock Nexus 5X, without any customization. Since it's not related to your PR, probably best to open a new issue for it - That toast was very barely noticeable in the context of this screen.

malinajirka commented:
We could just show a snackbar instead of a toast message I guess. Wdyt?

iamthomasbishop commented:
Snackbar seems appropriate here.

malinajirka commented:
Thanks! I fixed the issue with Travis and I created a new issue for Toast2Snackbar migration - https://github.com/wordpress-mobile/WordPress-Android/issues/8435

It's ready for another round. Thanks @daniloercoli

malinajirka commented:
@daniloercoli Can you please merge it, so I don't merge my own PR. Thanks!
